cl/citations/tasks.py

```diff
@@ -126,31 +132,35 @@ def find_citations_for_opinion_by_pks(self, opinion_pks, index=True):
         # values = number of times that opinion is cited
         grouped_matches = Counter(citation_matches)
 
+        # Increase citation count for each matched cluster if it hasn't
+        # already been cited by this opinion.
+        all_cited_opinions = opinion.opinions_cited.all()
         for matched_opinion in grouped_matches:
-            # Increase citation count for matched cluster if it hasn't
-            # already been cited by this opinion.
-            if matched_opinion not in opinion.opinions_cited.all():
-                matched_opinion.cluster.citation_count += 1
-                matched_opinion.cluster.save(index=index)
-
-        # Only update things if we found citations
-        if citations:
-            opinion.html_with_citations = create_cited_html(opinion, citations)
-
-            # Nuke existing citations
-            OpinionsCited.objects.filter(citing_opinion_id=opinion.pk).delete()
-
-            # Create the new ones.
-            OpinionsCited.objects.bulk_create(
-                [
-                    OpinionsCited(
-                        citing_opinion_id=opinion.pk,
-                        cited_opinion_id=matched_opinion.pk,
-                        depth=grouped_matches[matched_opinion],
-                    )
-                    for matched_opinion in grouped_matches
-                ]
-            )
+            if matched_opinion not in all_cited_opinions:
+                matched_opinion.cluster.citation_count = (
+                    F("citation_count") + 1
+                )
+                matched_opinion.cluster.save(
+                    index=index, update_fields=["citation_count"]
+                )
```

Comment:
OK, I did a little more research here, and I think we can actually make this a good deal faster than you already did. The trick is, like I was sort of thinking yesterday, doing fewer queries, though I didn't know exactly how yesterday.

So right now if we get a lot of matched citations in a given opinion, let's say there are 70 of them, we'll do 70 saves. Each creates a round trip to the DB, and possibly gets sent to solr if index=True, as a separate celery task. We can bring these numbers way down.

```
all_cited_opinions = opinion.opinions_cited.all().values_list('pk', flat=True)
opinion_ids_to_update = set()  # Keep a running list of things that need updating
for matched_opinion in grouped_matches:
    if matched_opinion.pk not in all_cited_opinions:
        opinion_ids_to_update.add(matched_opinion.pk)

# Increment all counters in one query
OpinionCluster.objects.filter(pk__in=opinion_ids_to_update).update(F("citation_count")+1)  # Fun, right?
if index:  # A single Solr update
    add_items_to_solr.delay(
        opinion_ids_to_update, "search.OpinionCluster",
    )
```

Reply:
Okay, made those changes in b77263d. I had to modify the `OpinionCluster.objects.filter(pk__in=opinion_ids_to_update)` lookup to be `OpinionCluster.objects.filter(sub_opinions__pk__in=opinion_ids_to_update)` instead -- does that look right to you?